### PR TITLE
CORE-9886: Enable Docker HUb Authentication against SSO  

### PR DESCRIPTION
**Pull request by knguyen-r3**
* Modified function setCredentialsOnBaseImage to be able to handle local Docker credentials to authenticate against DockerHub

### CODE REVIEW

**Comment by ronanbrowne** [comment]
FYI: I needed to force push to rebase this as old artifacts were outdated and could not be built locally.

**Comment by ronanbrowne** [comment]
@knguyen-r3 I had a look at this, and your issue is on line 220 

https://github.com/corda/corda-runtime-os/blob/knguyen/CORE-9886/enable_docker_authentication/buildSrc/src/main/groovy/DeployableContainerBuilder.groovy#L221

` builder = Jib.from(imageName)`
 
 is called, this is trying to pull the base image without authenticating.
 
Can you try and remove that line and replace it with and re test this.
  
` builder = setCredentialsOnBaseImage(builder)`

which will authenticate (presuming you have the relivent env vars set).

File: buildSrc/src/main/groovy/DeployableContainerBuilder.groovy [hunk 1]
```diff
@@ -326,11 +326,19 @@ abstract class DeployableContainerBuilder extends DefaultTask {
...
     private JibContainerBuilder setCredentialsOnBaseImage(JibContainerBuilder builder) {
+        logger.info("DEBUG setCredentialsOnBaseImage")
         def baseImage = RegistryImage.named("${baseImageName.get()}:${baseImageTag.get()}")
         if ((registryUsername.get() != null && !registryUsername.get().isEmpty()) && baseImageName.get().contains("software.r3.com")) {
+            logger.info("DEBUG artifactory setCredentialsOnBaseImage")
             baseImage.addCredential(registryUsername.get(), registryPassword.get())
             builder = Jib.from(baseImage)
         }
+        else if ((dockerHubUsername.get() != null && !dockerHubUsername.get().isEmpty()) && (dockerHubPassword.get() != null && !dockerHubPassword.get().isEmpty())){
+            logger.info("DEBUG setCredentialsOnBaseImage")
```
Comment (ronanbrowne) [inline buildSrc/src/main/groovy/DeployableContainerBuilder.groovy]:
```suggestion
            logger.info("Authenticating against Docker Hub for base image resolution")
```

File: buildSrc/src/main/groovy/DeployableContainerBuilder.groovy [hunk 2]
```diff
@@ -326,11 +326,19 @@ abstract class DeployableContainerBuilder extends DefaultTask {
      *  Set credentials on the base image we use
      */
     private JibContainerBuilder setCredentialsOnBaseImage(JibContainerBuilder builder) {
+        logger.info("DEBUG setCredentialsOnBaseImage")
         def baseImage = RegistryImage.named("${baseImageName.get()}:${baseImageTag.get()}")
         if ((registryUsername.get() != null && !registryUsername.get().isEmpty()) && baseImageName.get().contains("software.r3.com")) {
+            logger.info("DEBUG artifactory setCredentialsOnBaseImage")
```
Comment (ronanbrowne) [inline buildSrc/src/main/groovy/DeployableContainerBuilder.groovy]:
```suggestion
            logger.info("Authenticating against Artifactory for base image resolution")
```

File: buildSrc/src/main/groovy/DeployableContainerBuilder.groovy [hunk 3]
```diff
@@ -326,11 +326,19 @@ abstract class DeployableContainerBuilder extends DefaultTask {
...
             baseImage.addCredential(registryUsername.get(), registryPassword.get())
             builder = Jib.from(baseImage)
         }
+        else if ((dockerHubUsername.get() != null && !dockerHubUsername.get().isEmpty()) && (dockerHubPassword.get() != null && !dockerHubPassword.get().isEmpty())){
+            logger.info("DEBUG setCredentialsOnBaseImage")
+            baseImage.addCredential(dockerHubUsername.get(), dockerHubPassword.get())
+            builder = Jib.from(baseImage)
+        }
+        // Else need to authenticate locally on docker hub
```
Comment (ronanbrowne) [inline buildSrc/src/main/groovy/DeployableContainerBuilder.groovy]: I'd remove this comment as IMO it's unnecessary

File: buildSrc/src/main/groovy/DeployableContainerBuilder.groovy [hunk 4]
```diff
@@ -326,11 +326,19 @@ abstract class DeployableContainerBuilder extends DefaultTask {
      *  Set credentials on the base image we use
      */
     private JibContainerBuilder setCredentialsOnBaseImage(JibContainerBuilder builder) {
+        logger.info("DEBUG setCredentialsOnBaseImage")
```
Comment (ronanbrowne) [inline buildSrc/src/main/groovy/DeployableContainerBuilder.groovy]: remember to remove this line

**Review by ronanbrowne — CHANGES_REQUESTED** [review]
see comments  (i know still in draft)

**Review by ronanbrowne — APPROVED** [review]
LGTM

File: buildSrc/src/main/groovy/DeployableContainerBuilder.groovy [hunk 5]
```diff
@@ -328,9 +328,15 @@ abstract class DeployableContainerBuilder extends DefaultTask {
     private JibContainerBuilder setCredentialsOnBaseImage(JibContainerBuilder builder) {
         def baseImage = RegistryImage.named("${baseImageName.get()}:${baseImageTag.get()}")
         if ((registryUsername.get() != null && !registryUsername.get().isEmpty()) && baseImageName.get().contains("software.r3.com")) {
+            logger.info("Authenticating against Artifactory for base image resolution")
             baseImage.addCredential(registryUsername.get(), registryPassword.get())
             builder = Jib.from(baseImage)
         }
+        else if ((dockerHubUsername.get() != null && !dockerHubUsername.get().isEmpty()) && (dockerHubPassword.get() != null && !dockerHubPassword.get().isEmpty())){
```
Comment (yarivus) [inline buildSrc/src/main/groovy/DeployableContainerBuilder.groovy:335]: add a space before '{' and break the conditions to separate lines:
```
else if ((dockerHubUsername.get() != null && !dockerHubUsername.get().isEmpty()) &&
         (dockerHubPassword.get() != null && !dockerHubPassword.get().isEmpty())) {
```
See the previous 'if' and apply the same.

**Pull request merged**